broker/src/bin/neon_broker.rs

```diff
@@ -0,0 +1,563 @@
+        // channels.
+        let announce = match sub_key {
+            SubscriptionKey::All => {
+                assert!(self.subs_to_all.contains_key(&sub_id));
+                self.subs_to_all.remove(&sub_id);
+                NUM_SUBS_ALL.set(self.subs_to_all.len() as i64);
+                None
+            }
+            SubscriptionKey::Timeline(ref ttid) => {
+                remove_sub(&mut self.subs_to_timelines, ttid, sub_id);
+                self.num_subs_to_timelines -= 1;
+                NUM_SUBS_TIMELINE.set(self.num_subs_to_timelines);
+                Some(SubAnnounce::RemoveTimeline(*ttid, sub_id))
+            }
+        };
+        announce.map(|a| (self.pub_txs.values().cloned().collect(), a))
+    }
+}
+
+// SharedState wrapper for post-locking operations (sending to pub_tx chans).
+#[derive(Clone)]
+struct Registry {
+    shared_state: Arc<Mutex<SharedState>>,
+}
+
+impl Registry {
+    // Register new publisher in shared state.
+    pub fn register_publisher(&self) -> Publisher {
+        let (announce_tx, announce_rx) = mpsc::channel(128);
+        let mut ss = self.shared_state.lock().unwrap();
+        let id = ss.register_publisher(announce_tx);
+        let (subs_to_all, subs_to_timelines) = (
+            ss.subs_to_all.values().cloned().collect(),
+            ss.subs_to_timelines.clone(),
+        );
+        drop(ss);
+        trace!("registered publisher {}", id);
+        Publisher {
+            id,
+            announce_rx: announce_rx.into(),
+            subs_to_all,
+            subs_to_timelines,
+            registry: self.clone(),
+        }
+    }
+
+    pub fn unregister_publisher(&self, publisher: &Publisher) {
+        self.shared_state
+            .lock()
+            .unwrap()
+            .unregister_publisher(publisher.id);
+        trace!("unregistered publisher {}", publisher.id);
+    }
+
+    // Register new subscriber in shared state.
+    pub async fn register_subscriber(&self, sub_key: SubscriptionKey) -> Subscriber {
+        let (tx, rx) = mpsc::channel(get_args().chan_size);
+        let id;
+        let mut pub_txs;
+        let announce;
+        {
+            let mut ss = self.shared_state.lock().unwrap();
+            (id, pub_txs, announce) = ss.register_subscriber(sub_key, tx);
+        }
+        // Note: it is important to create Subscriber before .await. If client
+        // disconnects during await, which would terminate the Future we still
+        // need to run Subscriber's drop() which will unregister it from the
+        // shared state.
+        let subscriber = Subscriber {
+            id,
+            key: sub_key,
+            sub_rx: rx,
+            registry: self.clone(),
+        };
+        // Notify existing publishers about new subscriber.
+        for pub_tx in pub_txs.iter_mut() {
+            // Closed channel is fine; it means publisher has gone.
+            pub_tx.send(announce.clone()).await.ok();
```

Comment:
This and next `.await`'s are done in a loop on bounded (256 messages is the default size) channel that is being emptied by the receiver (if I've got that right, the receiver end goes into `while let Some(info) = subscriber.sub_rx.recv().await {`).


Since we don't guarantee what happens on the client side, there could be situations where a client opens connection and holds it more than 256 seconds due to some bug: maybe, it's slowly receiving WAL and spends resources there.

Then, every such await will wait for some time, not allowing the rest of the `pub_tx` to receive their announcements faster.

Instead, I think it's better to submit such tasks into `tokio::task::JoinSet` or `FuturesUnordered`/`FuturesOrdered` and .await on that concurrently.

-----

This is somewhat related to https://github.com/neondatabase/neon/pull/2733#discussion_r1012156020
since we don't really have to hold that many messages per channel for our tasks, IMO: we use `etcd` as a "blackboard", where we saw the latest metadata on the topic when we requested it, not necessarily seeing every update to the metadata.
In the old code, we did skip towards the latest version of the message per timeline in walreceiver, and only then applied some logic on top of the latest message.

Reply:
This particular send doesn't carry data, it just notifies publishers about the new subscriber (happens only on new subscription event).

The whole thing is designed not to block on overflow. If publisher can't send to subscriber, it drops the message. Add of new subs and drop of old ones is a rare operation, but anyway publisher is expected to consume these notifications quickly as this requires nothing but a bit of CPU.